modules/pam_userdb/pam_userdb.c

```diff
@@ -213,30 +213,39 @@ user_lookup (pam_handle_t *pamh, const char *database, const char *cryptmode,
 	  } else if (ctrl & PAM_ICASE_ARG) {
 	    compare = -2;
 	  } else {
+		char *setting = malloc(data.dsize+1);
+		if (setting != NULL) {
+			strncpy(setting, data.dptr, data.dsize);
+			setting[data.dsize] = '\0';
+			if (ctrl & PAM_DEBUG_ARG) {
+				pam_syslog(pamh, LOG_INFO, "crypt setting: %s", setting);
+			}
 #ifdef HAVE_CRYPT_R
 	    struct crypt_data *cdata = NULL;
 	    cdata = malloc(sizeof(*cdata));
 	    if (cdata != NULL) {
 		cdata->initialized = 0;
-		cryptpw = crypt_r(pass, data.dptr, cdata);
-	    }
+		cryptpw = crypt_r(pass, setting, cdata);
+		free(cdata);
+		}
 #else
-	    cryptpw = crypt (pass, data.dptr);
+	    cryptpw = crypt (pass, setting);
 #endif
+		free(setting);
+	    }
 	    if (cryptpw && strlen(cryptpw) == (size_t)data.dsize) {
 	      compare = memcmp(data.dptr, cryptpw, data.dsize);
 	    } else {
 	      compare = -2;
 	      if (ctrl & PAM_DEBUG_ARG) {
-		if (cryptpw)
+		if (cryptpw) {
 		  pam_syslog(pamh, LOG_INFO, "lengths of computed and stored hashes differ");
-		else
+		  pam_syslog(pamh, LOG_INFO, "computed hash: %s", cryptpw);
```

Comment:
I do not think we ever want the password hashes to appear in the syslog.

Reply:
Hmm, but we do print the hashes from the DB above with ctrl&PAM_DEBUG_ARG already. OK, then. But could you please amend the pam_userdb.8.xml so it prominently notes that with the debug option it will print password hashes from the db and computed from the entered password to the syslog?